### PR TITLE
P2P sanity tests #96

### PR DESCRIPTION
**Pull request by itsbalamurali**
This PR contains POC as defined in #96 by @moshababo.

### CODE REVIEW

**Comment by moshababo** [comment]
When running it on my local macOS env, it fails on `mkdirRecursiveSync`.
Regardless of how to solve this, and why it doesn't happen on Travis, I think we should avoid doing these kind of operations on the test file. We should adjust `xud` to support multiple instances, without requiring any hacks.
Take notice that if you apply `xudir` to the config args, it's being applied only after `config.load()` attempts to create it if it doesn't exists. So that flow doesn't seems to be well defined right now.

**Comment by itsbalamurali** [comment]
@moshababo I've fixed the root cause in `xud`. That error was happening since we are trying to create `xud` config directory ahead of overwriting the passed configuration arguments.

But for some reason one of the checks/assertions is failing, Can you help me with this?

File: package-lock.json [hunk 1]
```diff
@@ -3660,35 +3665,35 @@
...
-              "resolved": "https://registry.npmjs.org/minimist/-/minimist-0.0.8.tgz",
+              "resolved": false,
               "integrity": "sha1-hX/Kv8M5fSYluCKCYuhqp6ARsF0="
             }
           }
         },
         "ms": {
           "version": "2.0.0",
-          "resolved": "https://registry.npmjs.org/ms/-/ms-2.0.0.tgz",
+          "resolved": false,
```
[the remark this answers is not in the document]
Comment (itsbalamurali) [reply package-lock.json]: I am not sure about this, I always do `npm i` as soon as I pull the changes.

[the remark this answers is not in the document]
Comment (sangaman) [reply package-lock.json]: @itsbalamurali What version of npm are you using? Making sure you're on 6.x might help.

**Comment by itsbalamurali** [comment]
Looks we have a problem with `mysql2` package.

screen shot 2018-06-25 at 6 26 01 pm

**Comment by itsbalamurali** [comment]
Thanks, @sangaman  for the `mysql2` fix.

**Comment by moshababo** [comment]
It's still failing on `xud` startup (non-test), and the TOML file doesn't seem to override the config.

**Comment by itsbalamurali** [comment]
@moshababo  can you post a screenshot?

**Comment by moshababo** [comment]
Here's the error for when trying to launch `xud`.

screen shot 2018-06-28 at 20 11 03

The problem with the TOML file you could easily debug. Specifically in my case i'm getting an error because my mysql server requires a password, but that's not relevant to the problem.

**Comment by itsbalamurali** [comment]
@moshababo can you review once again and close this?

**Comment by moshababo** [comment]
My review is the same as it was.. nothing changes yet.

**Comment by moshababo** [comment]
My review is the same as it was.. nothing changed.

**Comment by moshababo** [comment]
In regards to the 2nd problem, this is my suggestion:
* don't rely on overriding `xudir`. it's being calculated dynamically, according to the OS platform.
* one single `xud.conf` file should be applied to all instances (the per-instance config are being specified in the test file anyway). 
* If any further distinction between different instances is necessary (`NodeKey`, logging, etc.), add support for instance id, and apply it.

**Comment by itsbalamurali** [comment]
@moshababo fixed issue and added support for instanceid you can notice the same in below screenshot.

screenshot 2018-07-21 at 12 49 38 am

**Comment by itsbalamurali** [comment]
@moshababo i am resolving the conflicts

File: lib/Logger.ts [hunk 2]
```diff
@@ -37,30 +37,27 @@ class Logger {
...
-  public static global = new Logger({ context: Context.GLOBAL });
-  public static db = new Logger({ context: Context.DB });
-  public static rpc = new Logger({ context: Context.RPC });
-  public static p2p = new Logger({ context: Context.P2P });
-  public static orderbook = new Logger({ context: Context.ORDERBOOK });
-  public static lnd = new Logger({ context: Context.LND });
-  public static raiden = new Logger({ context: Context.RAIDEN });
-
-  constructor({ level, logDir, context }: { level?: string, logDir?: string, context: Context}) {
+  constructor({ instanceId, level, logDir, context }: {instanceId?: number, level?: string, logDir?: string, context: Context}) {
```
[the remark this answers is not in the document]
Comment (itsbalamurali) [reply lib/Logger.ts]: Then how would i access these instance with them being static properties of Logger class when calling from other classes?

**Comment by itsbalamurali** [comment]
@moshababo  Then how would I access these instances with them being static properties of Logger class when calling from other classes?

**Comment by moshababo** [comment]
All the context instances will be on one main instance (you'll obviously need to change `Logger` to support that). You will pass them to all necessary classes, same as you previously did with `instanceId`, or instead pass them the main logger instance, and they will pick their desired context instance.

**Comment by itsbalamurali** [comment]
@moshababo please have a look at the error occurring due to recursion calls.

**Comment by moshababo** [comment]
Will probably get to it only on Sunday.

**Comment by moshababo** [comment]
You can't handle `Logger` both as a collection of loggers + a logger by itself. If the `Logger` instances within `Logger` are not static, it will lead to infinite recursion. You need another class to support that.

**Comment by moshababo** [comment]
Why are you passing `contextLogger` instead of just the specific logger?

**Comment by moshababo** [comment]
@itsbalamurali in case my last comment wasn't clear:
The general approach is to make the least possible changes to the code base in order to allow multiple xud instances support.

You created new class `ContextLogger` to aggregate the logger instances, which also could have been done with a simple helper function. At the moment there's no additional usage to this class, so instead of introducing it to all classes which are using a logger, you could simply just pass their specific context logger instance. The current situation of calling this class instance also a `logger` is a bit confusing.

**Comment by moshababo** [comment]
I've finished the open issue with the logger. a logger instance is now being passed to the constructor of all necessary classes. 

The tests are currently passing only on the first run. database need to be cleaned. this should be taken care of after #331 

The tests are very basic, and the interface `service.connect` need to improve (relying on enums/codes instead of strings). This should taken care of in our next p2p tests.

To support multiple xud instances run, changes were made throughout the codebase, mainly due to logging, but also to `NodeKey`. This PR was open for too long, so I won't wait for another review. We can open new issues if anything come up.

**Pull request merged**